airflow/models/taskreschedule.py

```diff
@@ -142,9 +154,93 @@ def find_for_task_instance(
         :param try_number: Look for TaskReschedule of the given try_number. Default is None which
             looks for the same try_number of the given task_instance.
         """
-        return TaskReschedule.query_for_task_instance(
-            task_instance, session=session, try_number=try_number
-        ).all()
+        return session.scalars(cls.stmt_for_task_instance(task_instance, try_number=try_number)).all()
+
+    @classmethod
+    @provide_session
+    def find_last_for_task_instance(
```

Comment:
I wonder if we really need this method at all. The logic is rather simple and does not really provide any encapsulation benefits. (Same for `find_for_task_instance` above but it’s already there so there’s probably an argument for keeping it.)

Reply:
I change logic a bit methods

New (all marked as private):
`stmt_for_task_instance` -> Query builder
`find_task_reschedule_for_ti` -> Return single task_reschedule record, first or last
`find_task_reschedules_for_ti` -> replacement for `find_for_task_instance`, not used in production code, only in tests

Deprecated Methods, for backward compatibility
`query_for_task_instance`
`find_for_task_instance`

In general only `find_task_reschedule_for_ti` and `stmt_for_task_instance` uses, all other are redundant, and in theory could be dropped. However `find_task_reschedules_for_ti` help me to find my mistake during tests

Reply:
But I we don’t really need `find_task_reschedule_for_ti` and `find_task_reschedules_for_ti` since those are just one liners that pass most of the arguments (except `session`) to `stmt_for_task_instance` anyway. We can just call `stmt_for_task_instance` and execute the statement as appropriated directly where needed.

Reply:
I've implement by this way due to difference between legacy Query API and new Statement Based (i don't know correct name): you don't have to provide a DB session for build query statement

We could directly use `stmt_for_task_instance` into
- airflow/models/taskinstance.py
- airflow/ti_deps/deps/ready_to_reschedule.py

Unfortunetly we can't do that in `airflow/sensors/base.py`, because we do not open session here, however it is not a showstopper, if applicable I could create private method in `BaseSensorOperator` which would open a session, execute query against `stmt_for_task_instance` and return start_date

Reply:
You don’t even need a private method, just use `with create_session()`, that’s what `@provide_session` does anyway.

Reply:
Nice! I just need to figure out how to deal with tests now, I'm not a fan of mocking DB execution if we could avoid it.

Reply:
I've just thought, are `find_task_reschedule_for_ti` / `find_task_reschedules_for_ti` would be required for AIP-44?

Reply:
Actually might be keep only `find_task_reschedule_for_ti` and use it again, WDYT?